src/Collider2D.cpp

```diff
@@ -0,0 +1,151 @@
+
+#include "Entity.h"
+#include "AK/AK.h"
+
+std::shared_ptr<Collider2D> Collider2D::create()
+{
+    auto col2d = std::make_shared<Collider2D>(AK::Badge<Collider2D> {}, ColliderType2D::Circle, 1);
+    return col2d;
+}
+
+std::shared_ptr<Collider2D> Collider2D::create(ColliderType2D collider_type, float radius)
+{
+    auto col2d = std::make_shared<Collider2D>(AK::Badge<Collider2D> {}, collider_type, radius);
+    return col2d;
+}
+
+std::shared_ptr<Collider2D> Collider2D::create(ColliderType2D collider_type, const glm::vec2& bounds_dimensions)
+{
+    auto col2d = std::make_shared<Collider2D>(AK::Badge<Collider2D> {}, collider_type, bounds_dimensions);
+    return col2d;
+}
+
+Collider2D::Collider2D(AK::Badge<Collider2D>, ColliderType2D collider_type, float radius) : m_radius(radius),
+m_collider_type(collider_type)
+{
+}
+
+Collider2D::Collider2D(AK::Badge<Collider2D>, ColliderType2D collider_type, const glm::vec2& bounds_dimensions) : m_bounds_dimensions(bounds_dimensions),
+m_collider_type(collider_type)
+{
+}
+
+void Collider2D::initialize()
+{
+    Component::initialize();
+    MainScene::get_instance()->emplace_collider(std::dynamic_pointer_cast<Collider2D>(shared_from_this()));
+}
+
+void Collider2D::start()
+{
+    Component::start();
+    set_can_tick(true);
+}
+
+std::string Collider2D::get_name() const
+{
+    std::string const name = typeid(decltype(*this)).name();
+    return name.substr(6);
+}
+
+void Collider2D::separate(glm::vec3 c1, glm::vec3 c2, float r1, float r2)
+{
+    // This is for circle x circle
+    glm::vec2 const c1_2d = AK::convert_3d_to_2d(c1);
+    glm::vec2 const c2_2d = AK::convert_3d_to_2d(c2);
+
+    glm::vec2 const separation_vector_2d = glm::normalize(c1_2d - c2_2d) * (r1 + r2 - glm::distance(c1_2d, c2_2d));
+
+    m_separation_vector = AK::convert_2d_to_3d(separation_vector_2d);
+    m_half_separation_vector = m_separation_vector * 0.5f;
+
+    glm::vec3 const new_position = entity->transform->get_local_position() + m_half_separation_vector;
+    entity->transform->set_local_position(new_position);
+}
+
+void Collider2D::separate(glm::vec2 c, float rad, glm::vec2 f, float l, float r, float t, float b, bool negate)
+{
+    // This is for rectangle x circle
+
+    float separation_direction = 1;
+    if (negate)
+        separation_direction = -1;
+
+    glm::vec2 separation_vector_2d = {};
+
+
+    if (c == f)
+    {
+        float const left = c.x - l + rad;
+        float const right = r - c.x + rad;
+        float const top = c.y - t + rad;
+        float const bottom = b - c.y + rad;
+
+        left < right ? separation_vector_2d.x = -left : separation_vector_2d.x = right;
+        top < bottom ? separation_vector_2d.y = -top : separation_vector_2d.y = bottom;
+
+        if (abs(separation_vector_2d.x) < abs(separation_vector_2d.y))
+            separation_vector_2d.y = 0;
+
+        if (abs(separation_vector_2d.x) > abs(separation_vector_2d.y))
+            separation_vector_2d.x = 0;
+    }
```

Comment:
Something is wrong. We never hit this path, and the value of glm::abs(c - f) looks like it will never be hit even if we use epsilonEqual

Reply:
> Something is wrong. We never hit this path, and the value of glm::abs(c - f) looks like it will never be hit even if we use epsilonEqual (unless it's big enough, then it will be hit almost always)

I have investigated this issue and although you haven't noticed any cases of hitting this path, this is correct.
That's just the case when **the center of the circle is located exactly on the rectangle**. This comes from the lecture: "Collisions in 2D, 2D Computer Games". The fact that when moving the objects around with ImGuizmo this condition is always false seems pretty intuitive then. Nevertheless, this doesn't mean that we shouldn't handle such case as it still may appear, eg. when spawning objects in the exact location where this case applies.